src/Kernel/RectorKernel.php

```diff
@@ -101,4 +119,97 @@ private function createDefaultConfigFiles(): array
     }
+
+    /**
+     * @param string[] $configFiles
+     */
+    private function createConfigsHash(array $configFiles): string
+    {
+        $fileHasher = new FileHasher();
+
+        if (self::$defaultFilesHash === null) {
+            self::$defaultFilesHash = $fileHasher->hashFiles($this->createDefaultConfigFiles());
+        }
+
+        Assert::allString($configFiles);
+        $configHash = $fileHasher->hashFiles($configFiles);
+
+        return self::$defaultFilesHash . $configHash;
+    }
+
+    /**
+     * @param string[] $configFiles
+     */
+    private function buildContainer(array $configFiles, string $hash = null): ContainerBuilder
+    {
+        $defaultConfigFiles = $this->createDefaultConfigFiles();
+        $configFiles = array_merge($defaultConfigFiles, $configFiles);
+
+        $compilerPasses = $this->createCompilerPasses();
+
+        $configureCallMergingLoaderFactory = new ConfigureCallMergingLoaderFactory($this->configureCallValuesCollector);
+        $containerBuilderFactory = new ContainerBuilderFactory($configureCallMergingLoaderFactory);
+
+        $containerBuilder = $containerBuilderFactory->create($configFiles, $compilerPasses);
+
+        // @see https://symfony.com/blog/new-in-symfony-4-4-dependency-injection-improvements-part-1
+        $containerBuilder->setParameter('container.dumper.inline_factories', true);
+        // to fix reincluding files again
+        $containerBuilder->setParameter('container.dumper.inline_class_loader', false);
+
+        $containerBuilder->compile();
+
+        return $this->container = $containerBuilder;
+    }
+
+    /**
+     * @param string[] $configFiles
+     */
+    private function buildCachedContainer(array $configFiles): ContainerInterface {
+        $hash = $this->createConfigsHash($configFiles);
+
+        $filesystem = new SmartFileSystem();
+        $className = 'RectorKernel'.$hash;
+        $file = sys_get_temp_dir() .'/rector/'. self::CACHE_KEY .'-'.$hash.'.php';
+
+        if (file_exists($file)) {
+            require_once $file;
+            $className = '\\'.__NAMESPACE__ .'\\'. $className;
+            $container = new $className();
+            if (!$container instanceof ContainerInterface) {
+                throw new ShouldNotHappenException();
+            }
+        } else {
+            $container = $this->buildContainer($configFiles, $hash);
+
+            $dumper = new PhpDumper($container);
+            $dumpedContainer = $dumper->dump([
+                'class' => $className,
+                'namespace' => __NAMESPACE__
+            ]);
+            if (!is_string($dumpedContainer)) {
+                throw new ShouldNotHappenException();
+            }
+
+            $filesystem->dumpFile($file, $dumpedContainer);
+        }
+
+        return $container;
+    }
+
+    static public function clearCache(): void {
+        $dir = sys_get_temp_dir();
+        if (!is_writable($dir)) {
+            return;
+        }
+
```

Comment:
`if (! $this->dumpFileCache) {` check seems also needed here:

```suggestion
        if (! $this->dumpFileCache) {
            return;
        }
```

as the cache container only in tests.

Reply:
Its a static method which cannot check the property. 

Also it checks for writability of the folders in questions. 

I don't see an issue right now

Reply:
And it makes sense the cache is cleared when the user asks for it. No matter its run via test

Reply:
Ok 👍